### PR TITLE
SQL: Add ProjectableFilterableTable to SegmentsTable.

### PR DESCRIPTION
**Pull request by gianm**
This allows us to skip serialization of expensive fields such as shard_spec, dimensions, metrics, and last_compaction_state, if those fields are not actually being queried.

### CODE REVIEW

**Review by georgew5656 — APPROVED** [review]
looks good to me, thanks for the change

File: sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemSchema.java [hunk 1]
```diff
@@ -241,7 +250,7 @@ public Map<String, Table> getTableMap()
   /**
    * This table contains row per segment from metadata store as well as served segments.
    */
-  static class SegmentsTable extends AbstractTable implements ScannableTable
+  static class SegmentsTable extends AbstractTable implements ScannableTable, ProjectableFilterableTable
```
Comment (abhishekrb19) [inline sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemSchema.java:253]: Do we still need to implement `ScannableTable`? I'd think the new `scan(DataContext, List<RexNode>, int[])` signature would come with empty filters and null projects if it's not present in the query

File: sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemSchema.java [hunk 2]
```diff
@@ -157,6 +161,11 @@ public class SystemSchema extends AbstractSchema
       .add("replication_factor", ColumnType.LONG)
       .build();
 
+  static final int SEGMENTS_SHARD_SPEC_INDEX = SEGMENTS_SIGNATURE.indexOf("shard_spec");
+  static final int SEGMENTS_DIMENSIONS_INDEX = SEGMENTS_SIGNATURE.indexOf("dimensions");
+  static final int SEGMENTS_METRICS_INDEX = SEGMENTS_SIGNATURE.indexOf("metrics");
+  static final int SEGMENTS_LAST_COMPACTION_STATE_INDEX = SEGMENTS_SIGNATURE.indexOf("last_compaction_state");
```
Comment (abhishekrb19) [inline sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemSchema.java:167]: nit: all of these can be `private static`

**Review by abhishekrb19 — APPROVED** [review]
🚀

File: sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemSchema.java (hunk 1 above)
Comment (gianm) [reply sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemSchema.java]: It seems it doesn't. I removed `ScannableTable`.

File: sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemSchema.java (hunk 2 above)
Comment (gianm) [reply sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemSchema.java]: I did a refactor that makes these unnecessary. The new logic moves `writeValueAsString` to `projectSegmentsRow`, and avoids both `ObjectMapper.writeValueAsString(o)` and `o.toString()` when `o` isn't being selected.

**Pull request merged**